### PR TITLE
[PM-1815] Include Member Decryption Type in Token Response

### PR DESCRIPTION
**Pull request by justindbaur**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

- Return `MemberDecryptionType` from SSO Config in token response. 
- Don't disrupt existing logins

## Code changes

* **src/Identity/IdentityServer/BaseRequestValidator.cs:** Made `ICurrentContext` a protected `get` only property so that it can be injected in the base class and still available to classes that inherit `BaseRequestValidator`.
* **src/Identity/IdentityServer/CustomTokenRequestValidator.cs:** Made it attempt to get sso config from any user that has an `organizationId` claim. If it exists, and the feature is enabled

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Identity/IdentityServer/CustomTokenRequestValidator.cs [hunk 1]
```diff
@@ -101,6 +108,20 @@ public async Task ValidateAsync(CustomTokenRequestValidationContext context)
...
+        // through here
+        var ssoConfigData = await GetSsoConfigurationDataAsync(context.Result.ValidatedRequest.Subject);
+
+        // You can't put this below the user.MasterPassword != null check because TDE users can still have a MasterPassword
+        // It's worth noting that CurrentContext here will build a user in LaunchDarkly that is anonymous but DOES belong
+        // to an organization. So we will not be able to turn this feature on for only a single user, only for an entire 
+        // organization at a time.
+        if (ssoConfigData != null && _featureService.IsEnabled(FeatureFlagKeys.TrustedDeviceEncryption, CurrentContext))
+        {
+            context.Result.CustomResponse["MemberDecryptionType"] = ssoConfigData.MemberDecryptionType;
```
Comment (jlf0dev) [inline src/Identity/IdentityServer/CustomTokenRequestValidator.cs:122]: If they do have a Master Password, don't we need to return that as a decryption option here?

Comment (justindbaur) [reply src/Identity/IdentityServer/CustomTokenRequestValidator.cs:122]: @jlf0dev Do you mean if they have a master password we force the decryption type to `MasterPassword` or we need to ALSO return whether they have a password here?

File: src/Identity/IdentityServer/CustomTokenRequestValidator.cs [hunk 2]
```diff
@@ -101,6 +109,20 @@ public async Task ValidateAsync(CustomTokenRequestValidationContext context)
...
+        // through here
+        var ssoConfigData = await GetSsoConfigurationDataAsync(context.Result.ValidatedRequest.Subject);
+
+        // You can't put this below the user.MasterPassword != null check because TDE users can still have a MasterPassword
+        // It's worth noting that CurrentContext here will build a user in LaunchDarkly that is anonymous but DOES belong
+        // to an organization. So we will not be able to turn this feature on for only a single user, only for an entire 
+        // organization at a time.
+        if (ssoConfigData != null && _featureService.IsEnabled(FeatureFlagKeys.TrustedDeviceEncryption, CurrentContext))
+        {
+            context.Result.CustomResponse["MemberDecryptionOptions"] = CreateMemberDecryptionOptions(ssoConfigData, user).ToList();
```
Comment (jlf0dev) [inline src/Identity/IdentityServer/CustomTokenRequestValidator.cs:123]: Might want to rename this to `UserDecryptionOptions` now

```suggestion
            context.Result.CustomResponse["UserDecryptionOptions"] = CreateUserDecryptionOptions(ssoConfigData, user).ToList();
```

File: src/Identity/IdentityServer/CustomTokenRequestValidator.cs [hunk 3]
```diff
@@ -164,4 +197,28 @@ public async Task ValidateAsync(CustomTokenRequestValidationContext context)
         context.Result.IsError = true;
         context.Result.CustomResponse = customResponse;
     }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    private static IEnumerable<UserDecryptionOption> CreateMemberDecryptionOptions(SsoConfigurationData ssoConfigurationData, User user)
```
Comment (jlf0dev) [inline src/Identity/IdentityServer/CustomTokenRequestValidator.cs:204]:
```suggestion
    private static IEnumerable<UserDecryptionOption> CreateUserDecryptionOptions(SsoConfigurationData ssoConfigurationData, User user)
```

File: test/Identity.IntegrationTest/Endpoints/IdentityServerSsoTests.cs [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+        }));
+
+        // Assert
+        // If the organization has selected TrustedDeviceEncryption but the user still has their master password
+        // they can decrypt with either option
+        Assert.Equal(StatusCodes.Status200OK, context.Response.StatusCode);
+        using var responseBody = await AssertHelper.AssertResponseTypeIs<JsonDocument>(context);
+        var root = responseBody.RootElement;
+        AssertHelper.AssertJsonProperty(root, "access_token", JsonValueKind.String);
+        var memberDecryptionOptions = AssertHelper.AssertJsonProperty(root, "MemberDecryptionOptions", JsonValueKind.Array).EnumerateArray();
```
Comment (jlf0dev) [inline test/Identity.IntegrationTest/Endpoints/IdentityServerSsoTests.cs:104]: Rename here as well

**Review by jlf0dev — COMMENTED** [review]
couple minor tweaks

File: src/Identity/IdentityServer/CustomTokenRequestValidator.cs (hunk 2 above)
Comment (justindbaur) [reply src/Identity/IdentityServer/CustomTokenRequestValidator.cs]: Good catch!

File: src/Identity/IdentityServer/CustomTokenRequestValidator.cs [hunk 5]
```diff
@@ -164,4 +197,28 @@ public async Task ValidateAsync(CustomTokenRequestValidationContext context)
         context.Result.IsError = true;
         context.Result.CustomResponse = customResponse;
     }
+
+    /// <summary>
+    /// 
+    /// </summary>
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply src/Identity/IdentityServer/CustomTokenRequestValidator.cs:203]: Added one

**Review by jlf0dev — APPROVED** [review]
LGTM

File: src/Core/Auth/Models/Api/Response/UserDecryptionOptions.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+public class UserDecryptionOptions : ResponseModel
+{
+    public UserDecryptionOptions() : base("userDecryptionOptions")
+    {
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
```
Comment (jlf0dev) [inline src/Core/Auth/Models/Api/Response/UserDecryptionOptions.cs:16]: Are these supposed to be empty?

Comment (justindbaur) [reply src/Core/Auth/Models/Api/Response/UserDecryptionOptions.cs:16]: Will do a follow-up PR.

**Pull request merged**